### ISSUE TITLE
Union types doesn't work as expected when I'm trying to use !!

### ISSUE THREAD

**Issue by mojtabast** [microsoft/TypeScript#32095]
**TypeScript Version:**  3.5.1

**Code**

```ts
interface DataType{

    display_name: string | null
}

function myName(data :DataType) : string{

    if( !!data.display_name === false ) return 'No Name'

    return data.display_name
}

```

**Expected behavior:**
Typescript shouldn't complain about ```return data.display_name``` because I covered the `null` before this.

**Actual behavior:**
It saying `Type 'string | null' is not assignable to type 'string'.
  Type 'null' is not assignable to type 'string'.` on ```return data.display_name``` 

**Playground Link:** 
[You can see the result here.](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgCJzHAKgTwA4QDeAUMcucgCbADOeANnDgPohwC2EAXMjWFKADmyAD7IQAV3r1iAX1IwJIBGGAB7EMnY4AchwgAKShjjIu6TLgIBKM735CSZCsBgHkAQg-HMAOmp0jCxsnMgAvBHI8PQ0KLZQEGASUJoA5DpqyHqcqaQUyAlJKVQm-rQMTKz6cqRAA)

**Comment by RyanCavanaugh** [comment]
Type guards only trigger on certain syntactic constructs, and there's no reason to write the form `if (!!x === false)` so TS doesn't recognize it.

**Comment by mojtabast** [comment]
I think this is a common way to coerce `null` & `undefined` & `string` to boolean. [This is one of the use cases that I faced with this issue](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgCJzHAKgTwA4QDeAUMcucgCbADOeANnDgPohwC2EAXMjWFKADmyAD7IQAV3r1iAX1IwJIBGGAB7EFTUBlNVDAAKerTA90mXAQDaAXQCUJMhSgQwEqJuN8AdDT2HFZVUNZDh6PAALOAAjV2AEMN19AzgAGmRou2RHClzkAHp85GAYUKoMOGQomnE1MCpaBiZWDgh0gHcUOBdefyFi+rA1ZDAIlAgQSmQ1Uq8wJzzkBA0+UIA5VuQAXlDvajpGFjZOBbzlkFXojc5tjL3Gw5aT09yS5AMAQg+4a5Qt-+Q8HoNAgWRcbg8yAAjC8KG9Ph8rpt-jsgSCwa53JoALQw2HkeE-TYAHgyvyyJEWi3BWOQuPxyHkDMJv2QAD4ya0KQzcjTIXiqYyGXzNAAGU6yOxyIA) 

I'm not sure about how TS can solve this, but I think this is a common way to write so for satisfying TS we need to add [some workaround like this](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgCJzHAKgTwA4QDeAUMcucgCbADOeANnDgPohwC2EAXMjWFKADmyAD7IQAV3r1iAX1IwJIBGGAB7EFTUBlNVDAAKerTA90mXAQDaAXQCUJMhSgQwEqJuN8AdDT2HFZVUNZDh6PAALOAAjV2AEMN19AzgAGmRou2RHClzkAHp85GAYUKoMOGQomnE1MCpaBiZWDgh0gHcUOBdefyFi+rA1ZDAIlAgQSmQ1Uq8wJzzkBA0+UIA5VuQAXlDvajpGFjZOBbzlkFXojc5tjL3Gw5aT4gBIRZLkFOuULd-xKXoojEAEI4N8si43B5kABGV7vUoGK6bX47STSIHIYHIzgQ1zuTQAWjhp1yHy+mwAPBlwdlSYtkJCCchifTyPI2cVEWDNgA+GmtLIkBl5JnQkki+QisWaAAMp1kdjkQA).

[I also checked `!val` is working as expected](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgCJzHAKgTwA4QDeAUMcucgCbADOeANnDgPohwC2EAXMjWFKADmyAD7IQAV3r1iAX1IwJIBGGAB7EFTUBlNVDAAKerTA90mXAQDaAXQCUJMhSgQwEqJuN8AdDT2HFZVUNZDh6PAALOAAjV2AEMN19AzgAGmRou2RHClzkAHp85GAYUKoMOGQomnE1MCpaBiZWDgh0gHcUOBdefyFi+rA1ZDAIlAgQSmQ1Uq8wJzzkBA0+UIA5VuQAXlDvajpGFjZOBbzlkFXojc5tjL3Gw5aT09yS5AMAQjhriCyXNw8yAAjC8KG9PldWn9XO5NABaEGg8jg76bAA8GR+WRIi0W-1hyARSOQ8mJKJ+yAAfJiodlibl8YDEbiScTGZoAAynWR2ORAA) but `!!` is not working as I said.

**Comment by nattthebear** [comment]
`!!` is a common way to coerce to boolean, and Typescript does recognize it for narrowing:

https://www.typescriptlang.org/play/index.html#code/CYUwxgNghgTiAEA3W8oC54G94A8MBZ4BfeAH3gDsBXCCAKDoEsAzeACgEIOoBKLO+INQBueAHoxWXAWJ0SICAGcEmAUKiiJlGvSJ0gA

Putting `=== false` or `=== true` in an `if` is the thing that nobody does.

**Comment by mojtabast** [comment]
@nattthebear  So maybe I'm the only one writes in this way :)   

In your example, if we put `a` outside of the else it doesn't work properly in function at least. [see this](https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAcghgWwgFXNAvFAzsATgSwDsBzKAHykIFcAbGgKHoDMrCBjYfAe0KgFkQ8JAApCALliIUaAJQBvRlCVR8TYQEJ1hefWV6ouCMCq5ehXcoC+i5YeOnK9S0A) 

Anyways my point is something that is working in JS, expected to work on TS too. this is a valid case IMO.

**Comment by RyanCavanaugh** [comment]
There's no limit to how elaborate you could make your condition, but TS pays a performance and complexity cost for every form it supports, and there's just no reason to write `!!x === false` instead of `!x`. If this were idiomatic and common we would consider it, but it appears to be atypical (i.e. you are the first person to notice).

**Comment by nattthebear** [comment]
Your second link gets into an unrelated question:  falsy checks on a string don't remove string from the possibilities because `""` is falsy.  This has nothing to do with the `=== false` construct and TS is behaving correctly.
